vllm/spec_decode/spec_decode_worker.py

```diff
@@ -195,6 +192,19 @@ def __init__(
         self._metrics = AsyncMetricsCollector(
             self.spec_decode_sampler
         ) if metrics_collector is None else metrics_collector
+        # Tracks the sequence IDs that received a bonus token ID in
+        # their last forward pass. Needed only if KV cache is being
+        # used for token generation such as in the case of MultiStepWorker.
+        self.seq_with_bonus_token_in_last_step: Set[int] = set()
+        # Tracks the currently active request ids and the sequence IDs
+        # corresponding to them
+        self.request_id_seq_id_mapping: Dict[str, Set[int]] = defaultdict(set)
+        # Tracks if the proposer worker uses the KV cache or not.
+        self.proposer_uses_kv_cache = True
+        if (isinstance(self.proposer_worker,
+                       (MLPSpeculatorWorker, NGramWorker))):
+            self.proposer_uses_kv_cache = False
```

Comment:
let's move this logic to `create_worker`? we want to minimize data-dependent control flow in constructor much as possible so testing is easier

Reply:
I removed this completely. We are always maintaining the bonus token related data structures even if it might not be needed by the worker.